File: backends/java-spring/src/main/java/com/authorization/infrastructure/cache/RedisGateRepository.java
/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */
package com.authorization.infrastructure.cache;

import com.authorization.config.AppConfig;
import com.authorization.domain.gate.repositories.GateRepository;
import org.springframework.stereotype.Component;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;

/**
 * Redis を使用したゲートキャッシュRepositoryの実装です。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
@Component
public class RedisGateRepository implements GateRepository {

    private final JedisPool pool;
    private final String cachePrefix;

    public RedisGateRepository(JedisPool pool, AppConfig cfg) {
        this.pool = pool;
        this.cachePrefix = cfg.app().cachePrefix();
    }

    @Override
    public void putJwt(String identifier, String memberId, String token, int ttl) {
        try (Jedis jedis = pool.getResource()) {
            jedis.setex(cacheKey(identifier, memberId), ttl, token);
        }
    }

    @Override
    public String getJwt(String identifier, String memberId) {
        try (Jedis jedis = pool.getResource()) {
            return jedis.get(cacheKey(identifier, memberId));
        }
    }

    /**
     * キャッシュキーを組み立てます。
     *
     * @param identifier クライアント識別子
     * @param memberId メンバーID
     * @return キャッシュキー
     */
    private String cacheKey(String identifier, String memberId) {
        return cachePrefix + ":gate.jwt:" + identifier + ":" + memberId;
    }
}

File: backends/kotlin-ktor/src/main/kotlin/com/authorization/infrastructure/cache/RedisGateRepository.kt
/*
 * ゲートキャッシュリポジトリの Redis 実装モジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
package com.authorization.infrastructure.cache

import com.authorization.config.Config
import com.authorization.domain.gate.CacheRepository
import redis.clients.jedis.JedisPool

/**
 * Redis を使用したゲートキャッシュリポジトリの実装です。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
class RedisGateRepository(
    private val pool: JedisPool,
    private val cfg: Config,
) : CacheRepository {

    /**
     * キャッシュから JWT を取得します。
     *
     * @param identifier クライアント識別子
     * @param memberId メンバー ID
     * @return JWT、またはキャッシュがなければ null
     */
    override suspend fun getJwt(identifier: String, memberId: String): String? = TODO()

    /**
     * JWT をキャッシュに保存します。
     *
     * @param identifier クライアント識別子
     * @param memberId メンバー ID
     * @param token JWT
     * @param ttl キャッシュ有効期間（秒）
     */
    override suspend fun putJwt(identifier: String, memberId: String, token: String, ttl: Long) = TODO()
}
